firmware/tools/bitmap_tools/pp_png2hpp.py

```diff
@@ -0,0 +1,217 @@
+	bmp_data.append(f"""}};
+static constexpr Bitmap bitmap_{name}{{
+    {{{str(rgb_im.size[0])}, {str(rgb_im.size[1])}}},
+    bitmap_{name}_data}};
+""")
+	
+	out_bmpdata = ''.join(map(str, bmp_data))
+	
+	return (out_bmpdata)
+
+def pp_bitmaphpp_data(pngicons_path):
+    count = 0
+    bitmaphpp_data = []
+    for file in os.listdir(pngicons_path):
+        if os.path.isfile(pngicons_path + file) and file.endswith(".png"):
+            bitmaphpp_data.append(convert_png(pngicons_path + file))
+            count += 1
+
+    return bitmaphpp_data
+    
+
+def pp_bitmaphpp_footer():
+    return("""
+} /* namespace ui */
+	
+#endif /*__BITMAP_HPP__*/
+""")
+
+
+def pp_write_bitmaphpp(pngicons_path, hpp_outpath):
+    bitmaphpp_file = []
+
+    bitmaphpp_file.append(pp_bitmaphpp_header())
+    bitmaphpp_file.append("".join(str(x) for x in pp_bitmaphpp_data(pngicons_path)))
+    bitmaphpp_file.append(pp_bitmaphpp_footer())
+    
+    out_file = "".join(str(x) for x in bitmaphpp_file)
+
+    with open(hpp_outpath, "w", encoding="utf-8") as fd:
+        fd.writelines(out_file)
+    
+    print("Find your bitmap.hpp at", out_file)
+
+
+### Convert from a bitmap.hpp file one or all icons in png.
+###########################################################
+
+def parse_bitmaphpp(bitmaphpp_file):
+    ico_pattern = re.compile(r"static constexpr uint8_t bitmap_(.*)_data\[\] = {\n((?:\s+(?:.*)\n)+)};\nstatic constexpr Bitmap bitmap_.*\{\n\s+\{(.*)\},", re.MULTILINE)
+    ico_data = []
+    
+    # read file to buffer, to find multiline regex
+    readfile = open(bitmaphpp_file,'r')
+    buff = readfile.read()
+    readfile.close()
+
+    for match in ico_pattern.finditer(buff):
+        ico_data.append([match.group(1), match.group(2), match.group(3)])
+
+    return (ico_data)
+
+
+def convert_hpp(icon_name,bitmap_array,iconsize_str,png_outdir):
+    iconsize = iconsize_str.split(", ")
+    bitmap_size = (int(iconsize[0]),int(iconsize[1]))
+    bitmap_data=[]
+
+    image_data = np.zeros((bitmap_size[1], bitmap_size[0]), dtype=np.uint8)
+
+    for value in bitmap_array.split(",\n"):
+        if (value):
+            bitmap_data.append(int(value, 0))
+
+    for y in range(bitmap_size[1]):
+        for x in range(bitmap_size[0]):
+            byte_index = (y * bitmap_size[0] + x) // 8
+            bit_index = x % 8
+            # bit_index = 7 - (x % 8)
+            pixel_value = (bitmap_data[byte_index] >> bit_index) & 1
+            image_data[y, x] = pixel_value * 255
+
+    image = Image.fromarray(image_data, 'L')
+    icon_out = os.path.join(png_outdir,icon_name+".png")
+    image.save(icon_out)
+
+
+### Processing
+##############
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument("hpp", help="Path for bitmap.hpp")
+    parser.add_argument("graphics", help="Path of <icon>.png files to convert", default=os.path.join(os.getcwd(),"graphics"))
+    parser.add_argument("--icon","-i", help="Name of the icon to convert")
```

Comment:
It seems no code to handle this argument yet, if it's not yet finished, i think it's all good, we can just comment this argument out

Reply:
Just implemented. In future I will uncomment unused code.